### ISSUE TITLE
Support for GitHub Packages

### ISSUE THREAD

**Issue by cdhunt** [PowerShell/PSResourceGet#163]
I haven't found a way to use PowerShellGet 2.x to publish to [GitHub Packages](https://help.github.com/en/packages). Please include Packages as a supported repository provider.

All permutations of passing credentials (PAT) return a 401 (Unauthorized).

- Credential parameter
  - email and PAT
  - Org name and PAT
- NuGetApiKey parameter
  - email:PAT
  - Orgname:PAT
  - PAT

**Comment by JustinGrote** [comment]
I think Beta6 that was just released added publishing, and Github Packages is Nuget V3 so it may work, have you tried with beta6 yet?

**Comment by cdhunt** [comment]
I tried but ended up creating #172.

**Comment by jachin84** [comment]
I got past authentication by doing the following:

1.  Register GitHub Packages with nuget: `nuget.exe sources add -Name GitHub -Source https://nuget.pkg.github.com/ORGNAMEORUSERNAME/index.json -Username ORGNAMEORUSERNAME -Password <PAT>`
2. Register PSRepository: `Register-PSRepository -Name https://nuget.pkg.github.com/ORGNAMEORUSERNAME/index.json -SourceLocation $source -PublishLocation https://nuget.pkg.github.com/ORGNAMEORUSERNAME/index.json`
3. Attempt to publish: `Publish-Module -Path .\ModuleName -Repository GitHub -NuGetApiKey "n/a"` 

This gives the error `Failed to publish module 'ModuleName': 'nuget.exe failed to push Response status code does not indicate success: 400 (Bad Request). '`.

If you trace with Fiddler one of the headers give a clue:
>X-Nuget-Warning: No destination repository detected. Ensure the source project has a 'RepositoryUrl' property defined. If you're using a nuspec file, ensure that it has a repository element with the required 'type' and 'url' attributes.

GitHub Packages requires the RepositoryUrl element in the nuspec.
`<repository type="git" url="https://github.com/org/project" commit="50af1b9788e251964deb511745dd4716beee4e2b" />`

**Comment by IlyaVassyutovich** [comment]
I have the same need to be able to publish pwsh modules into GHPackages.

And I can confirm @jachin84 findings, GHPackages feed requires `<repository>`-configuration inside `.nuspec`. Unfortunately right now it is not supported by [`PublishPSResource.createNuspec`](https://github.com/PowerShell/PowerShellGet/blob/f1ddcb3bc602c61f9b98f79b1ab02b00331da8f5/src/code/PublishPSResource.cs#L507) and by `psd1`-format.

Now I'll try to create pull request to fix this, after figuring out how to use this repo for development.

### UPD
Oh...
>  As a result, we are currently not accepting PRs to this repository.

Ok, I think I'll stick with own fork of PowerShellGet2. But nonetheless, let's treat this as feature request? That is, ability to specify module repository metadata. 🤗

**Comment by cdhunt** [comment]
I was able to get Publish and Install working today. Publish was a three-step process.

1. `Register-PSResourceRepository -Name nuget-local -URL C:\Nuget\publish\`
1. `Publish-PSResource -Path .\mymodule\ -Repository nuget-local`
1. `gpr push -k <PAT> C:\Nuget\publish\mymodule.0.1.0.nupkg -r https://github.com/ORG/Repo`

[gpr](https://github.com/jcansdale/gpr) is a helper tool for pushing Nuget packages to GitHub Packages that makes it easy to provide the RepositoryURL metadata that GitHub Packages requires. You could also edit the nuspec inside the package after doing the local publish.

To install use [PowerShellGet 3.0.0-beta10](https://www.powershellgallery.com/packages/PowerShellGet/3.0.0-beta10):
1. `Register-PSResourceRepository -Name github -URL https://nuget.pkg.github.com/ORG/index.json`
1. `Install-PSResource -Name mymodule -Repository github -Credential (Get-Credential)`
  - Use Username and PAT for credential

**Comment by cdhunt** [comment]
Another hiccup. If your private package depends on a public package, PowerShellGet doesn't know how to find it since GitHub Packages doesn't support feed proxies like Proget and Artifactory. It looks for all of the dependent packages in GitHub and errors out if they can't be found.

**Comment by SydneyhSmith** [comment]
Thanks @cdhunt can you please clarify the relationship you expect from GitHub packages and Proget/Artifactory...is the package in GitHub and depends on resources in Proget/Artifactory?

**Comment by cdhunt** [comment]
@SydneyhSmith I wasn't expecting a relationship, just stating that GitHub Packages doesn't have the same repository proxy functionality as Proget and Artifactory where you can combine multiple feeds into one API endpoint. The result of that is if you define a module dependency in a private module hosted on GitHub Packages, Install-PSResource assumes the dependent package is also hosted on the same Repository and can't find it unless you publish a copy into your private repository.

In other words, I don't know of a mechanism to define module dependencies that come from different repositories with the current way they are defined in the module manifest. This is really a separate, but related issue I've been meaning to open an issue for.

**Comment by alerickson** [comment]
@cdhunt it would be great if GitHub could address this issue.  I do have some ideas of how we can address this client side, but because this isn't a prioritized scenario it likely wouldn't get addressed until Fall.  That said if it's an issue that's a major pain point for a lot of folks we can reconsider.

**Comment by JustinGrote** [comment]
I had mentioned that the dependency handling is kind of bespoke to powershell now and that it should leverage the already-established libraries for dependencies resolution in the nuget client unless theres a *really* strong reason not to.

**Comment by cdhunt** [comment]
@alerickson I'm.... okay... with the way GitHub Packages works. It does what we need, but without PowerShellGet client support, we have to do hacky things like copy public modules into a private feed or do dynamic install on Import-Module.

To the original issue, we have the same publish issues with GitHub Packages for other Nuget packages as well. `RepositoryUrl` is not part of the Nuget spec, but I don't see how to get around that. It's a requirement of how Packages is designed because feeds are global to an Org. So, unfortunately, I think all Nuget clients are going to have to adapt.

I think there are plenty of people who would be willing to work on that fix whenever you start accepting PRs, but it'll likely be a GitHub shim.

**Comment by brettmillerb** [comment]
Is there a v3 feed support roadmap in one of the existing milestones for GA or do we still need to jump through the above hoops outlined by @cdhunt for the time being @SydneyhSmith ?

I'm looking to implement some new internal tooling baselines but I'm hoping for native support before I start the work

**Comment by alerickson** [comment]
This is resolved in PR #1167
